### DISCUSSION TITLE
3.18.15 x 4.0

### DISCUSSION THREAD

**Discussion by paulobreim** [sepinf-inc/IPED#1226] [Q&A]
Is anyone having this problem below?
I indexed the same file in versions 3.18.15 and 4.0.1.
After indexing, I searched for the word "renova".
In version 3.18.15 it returned many files with a score of 100%, which are correct.
In version 4.0.1 only 3 occurrences have some percentage, but far from 100%.
Another information is that in version 4.0.1, the "default" did not bring the score in descending order as in 3.18.15.

Any explanation about this?

![Screenshot_20220721_195109](https://user-images.githubusercontent.com/381594/180328736-fec0623f-faec-4fa4-91ca-6fc5261fe683.png)

tks

**Comment by lfcnassif** [comment]
Hi @paulobreim,

The score absolute value itself, for index searches, means nothing. We updated the lucene indexing library from version 5.x to version 9.x, updating the score computation formula together, so absolute value changes are expected. The only rules it must follow is to give higher scores for hits on file names and for files with more hits. Ignoring hits on file names, the higher the number of hits in file content the higher the score should be. If that is not being respected, please let me know.

> Another information is that in version 4.0.1, the "default" did not bring the score in descending order as in 3.18.15.

This was intentional. Some internal changes in lucene library have broken some modules when the results are transversed in non index-id order (like in score order). To avoid having to re-sort the results (potentially many times) before calling the affected modules/features, they are being returned in index-id order now. If user has some column sorting applied (different than score), what I think is a very common scenario, returning results in score order is useless, because the column sorting will be applied. If you want to sort by score, simply click on the Score column header, as I have already suggested.

**Reply by paulobreim** [reply]
Tks.
I will do more tests.

**Reply by paulobreim** [reply]
I did some more tests and found the following:
In version 3.18.15, the score presented has 100% in a large number of files, but the information is not correct, because it assigns 100% to files that contain 64 occurrences of the searched word, but also assigns 100% for files that contain only 5 occurrences
The table below shows the number of occurrences and the assigned score.
1 - 47%
2 - 67%
3 - 97%
4 - 100%
5 - 100%
7 - 100%
10 - 100%
The same search in version 4.0.1 shows the following:
1 - 1%
2 - 2%
3 - 3%
4 - 4%
5 - 5%
6 - 6%
7 - 7%
8 - 8%
10 - 11%
16 - 16%
64 - 64%
The maximum result was 64%. Apparently the number of occurrences is equal to the Score.
So I did another search with a more common word and the result was:
2 - 2%
4 - 4%
5 - 5%
11/10 - 12%
12 - 14%
17 - 17%
23 - 23%
43 - 50%
57 - 70%
77 - 87%
83 - 97%
94 - 97%
169 - 100%
327 - 100%

Just even analyzing the code to try to see why this behavior.
---
I also noticed some situations where the counter is incorrect.
In the example below, 5 occurrences are shown, but the right one can observe that there are 6.

![Screenshot_20220722_182909](https://user-images.githubusercontent.com/381594/180573030-7ed10a48-f8a0-48ca-bbab-35c7c8fbcb97.png)

**Reply by lfcnassif** [reply]
Actually, many years ago, with old lucene 4.x, I tried to show the number of hits instead of  "score" but I wasn't able to make it work with phrase queries (2 or more terms between double quotes, try that) then. So I changed the lucene score formula to show some score "kind" proportional to the number of hits, since default lucene formula divides by content size to give higher scores to small docs with the same number of hits than large docs. So current intention is not to give the number of hits, but to show some score kind proportional to it. If you manage to compute the exact number of hits for all kinds of queries (look at the search help for them) let me know and I will happily apply your change.

About the hits panel, actually the title string is wrong, but the count should be ok. The current intended behavior is to count the number of text fragments with hits, not the number of hits. Take a look at it, I think there should be 1 text fragment with 2 hits in the same text chunk. The occurrences panel title should be fixed indeed.

**Reply by lfcnassif** [reply]
Maybe with current lucene 9 version, computing the exact number of hits for all query types could be easier, I'll try to take a look when I return from vacation.

**Reply by paulobreim** [reply]
I'll give my point of view on the score, but the way it is today, I think it's good.
This score theme is not easy to work with.
I did some more tests to think of something and here comes the question. How can the score help the investigator in a given objective? It may be that not everyone has the same thought, but logic says that the higher the score, the greater the probability that the file has what it is looking for, but we also know that a file that has only one occurrence of the search could be exactly what it is looking for. the investigator looks for.
There are even situations where I can find a file that has a large number of occurrences that causes the others to have a very low score.
That's why it's so difficult to establish a criterion, because all the files returned in the search will have to be examined.
So, from that point of view, maybe one way would be to consider occurrences between one and a hundred (above a hundred would be considered a hundred) and the score would be that number of occurrences.
The one currently shown by version 4.0.1 seems to be closer to this.
Enjoy the holidays and see it on the way back.

**Reply by lfcnassif** [reply]
I completely agree with you. Lucene tries to give the "best" hits using some score computation, but in real world cases, that is a very difficult task. So I simplified lucene formula to give a score proportional to the number of hits, at least it is a more objective value and easier to be understood by users.
